Rewrite unlinkability considerations (#354)
oauth-wg/oauth-selective-disclosure-jwt: draft-ietf-oauth-selective-disclosure-jwt.md

## Unlinkability

Unlinkability is a property that prevents an adversary from associating two or

> more sessions of the same user. Two types of unlinkability are often considered:

Reviewer: Is it really sessions..?

Reviewer 2: The privacy pass architecture uses this language:

>  These proofs, or tokens, are
>  private in the sense that a given token cannot be linked to the
>  protocol interaction where that token was initially issued.

> more sessions of the same user. Two types of unlinkability are often considered:

Reviewer: Is it really sessions..?

> Unlinkability is a property that prevents an adversary from associating two or
> more sessions of the same user. Two types of unlinkability are often considered:

Author: You're right, 'sessions' is not a concept we used previously. How about this?
```suggestion
Unlinkability is a property that prevents an adversary from associating credential
presentations in certain ways. Two types of unlinkability are often considered:
```

Reviewer: `from associating credential presentations of the same user to know about the behavior of that user beyond that user's consent`?
Probably some sentence is needed on why associating is a problem - ie adversary being able to know more about the user than it is supposed to, I think?

Reviewer 2: imo the threat is about distinguishability, it has to do with the anonymity set of the data subject... You might find t-closenss / or similar concepts helpful. https://en.wikipedia.org/wiki/T-closeness

Issuer / Verifier: Issuers are distinguishable, Subjects are distinguishable (signature correlation), Holders are not distinguishable (batch issuance).

Verifier / Verifier: Issuers are distinguishable, Subjects are not distinguishable (linked secret), Holders are not distinguishable (zkp).

In all cases, we assume multiple verifiers process presentations and are compromised, and then we evaluate the damage done to data subjects.

Reviewer 2: .... what I said above is not right... I think "Issuer / Verifier" unlinkability is a bad name... I think the security property we care about there is timing, metadata analysis.

Reviewer 2: ![Screen Shot 2023-10-30 at 1 53 04 PM](https://github.com/oauth-wg/oauth-selective-disclosure-jwt/assets/8295856/9f03157d-d735-4012-bafb-9338f9a96cd4)

Not sure if this helps, but the reason Issuer / Verifier Unlinkability is not achievable is the signature on the credential... batch issuance does not solve this, it just makes the attacker count unique signatures that are observed in issuers and verifiers.

Reviewer 2: I think these labels are possibly the wrong words to be using especially because "verifier / verifier" unlinkability seems stronger than issuer / verifier.... but is actually not.

Reviewer 2: Its possible that "single use unlinkability" and "multiple use unlinkability" might be better.

Author: @OR13  I'll make a bigger change to this PR hopefully addressing some of your comments, but certainly not everything. I propose that we try to get to an "acceptable" description first, get that merged, and then dig into the details in a separate issue.

- Verifier/Verifier unlinkability: Two colluding Verifiers should not be able to
  tell that they have received presentations from the same user.
- Issuer/Verifier unlinkability: An Issuer colluding with a Verifier should

>   not able to tell that a user that has received a credential from the Issuer

Reviewer 3: ```suggestion
  not be able to tell that a user that has received a credential from the Issuer
```

> - Issuer/Verifier unlinkability: An Issuer colluding with a Verifier should
>   not be able to tell that a user that has received a credential from the Issuer
>   has presented it to the Verifier.

Reviewer: ```suggestion
- Issuer/Verifier unlinkability: An Issuer colluding with a Verifier should
  not be able to tell to which Verifier a user has presented a credential issued from this Issuer.
```

Author: In the proposed text, the Verifier must collude with multiple Verifiers. I think my original definition is stronger.

Reviewer 4: Issuer/Verifier unlinkability: 

The basic property should not include the wording "colluding with a Verifier".

Proposal: An Issuer should not be able to know to which Verifier data derived from a credential issued by the Issuer has been presented to the Verifier.

Reviewer 2: mechanically I think there are 2 things that are trying to fit into a single sentence about "Issuer/Verifier unlinkability" here:

1. phone home / issuer's learning when a specific credential is presented.
2. tracking / fingerprinting of the subject by the issuer.

The first is mitigated partially by heard privacy.
The second requires the issuer not to assign or hide correlation capable identifiers.... and even the signature is such an identifier.

I suggest starting by just writing out the behaviors you want to see, and don't want to see... and then don't give the same name to multiple behaviors.

Author: The difference is effectively whether the Verifier is cooperating with the attacker or whether it is just following protocol; I'll split this up accordingly.

Reviewer 5: hey Dan, I don't love "to tell" and the concept of the speacking machines :-)

WDYT about this

````
Issuer/Verifier unlinkability: If an Issuer and a Verifier were to collude, they should not be able to determine that a user has presented a credential from the Issuer to that Verifier.
```

In both cases, unlinkability is limited to cases where the

> disclosed claims do not contain information that directly or indirectly

Reviewer: Really? There can be multiple same people with the same birthday. In which case, if the holder discloses only the birthdate to the verifier, the issuer/verifier unlinkability is possible.

Author: Yes, but in that case, the information was obviously not enough to identify the user.

Reviewer: I think your argument is that the example I give above still does not prevent issuer/verifier unlinkability in sd-jwt because the user's public key becomes the correlation factor, which I think needs to be explicitly stated

Author: No, what I want to say is: The amount of information that can identify a user varies depending on context knowledge and the set of users that exist. If I'm looking at a club with 12 members all having a different birthdates, the birthdate is sufficient identifying information. If there would be a specific credential for sitting US presidents, unlinkability cannot be achieved at all, even if only the credential type is shared.

So the point is: For batch issuance, even if all keys and so on are fresh for each credential, there is an upper limit on the grade of unlinkability achievable, and it depends on the dataset shared and context knowledge etc.

identifies the user. For example, when a tax ID is contained in the disclosed
claims, the Issuer and Verifier can easily link the user's sessions. However,
when the user only discloses a birthdate to one Verifier and a postal code to
another Verifier, the two Verifiers should not be able to tell that they were
talking to the same user.

> With SD-JWT, Issuer/Verifier unlinkablility cannot be achieved, as the issued

Reviewer 2: ```suggestion
Issuer/Verifier unlinkability cannot be achieved with SD-JWT, as the issued
```

> Issuer/Verifier unlinkability cannot be achieved with SD-JWT, as the issued
> SD-JWT is directly forwarded to the Verifier.

Reviewer: I don't understand this. Three party model in itself allows to realize issuer/verifier unlinkability. Are you trying to say that when the issuer directly forwards to the verifier it is not achieved? But in that case no need for sd-jwt.

```suggestion
```

Author: The three party model does not realize issuer/verifier unlinkability (per the definition above, it assumes colluding issuers and verifiers). This is one of the major differences between salted-hash based approaches and other approaches using more advanced cryptography.

> Issuer/Verifier unlinkability cannot be achieved with SD-JWT, as the issued
> SD-JWT is directly forwarded to the Verifier.

Author: ```suggestion
Issuer/Verifier unlinkability cannot be achieved in salted-hash based approaches to credentials, such as SD-JWT, as the issued
credential is directly forwarded to the Verifier.
```

> Verifier/Verifier unlinkablility can be achieved using batch issuance: A batch

Reviewer: ```suggestion
Verifier/Verifier unlinkablility can be achieved using batch issuance. A batch
```

Author: I'd like to keep this as-is.

## Unlinkability

> Unlinkability is a property whereby adversaries are prevented from correlating
> credential presentations of the same user (beyond the user's consent).

Reviewer 5: ```suggestion
Unlinkability is a property that prevents adversaries from correlating credential presentations of the same user beyond the user's consent.
```

Reviewer 5: likewise, dont worry and go ahead while you can :-)

Otherwise, an adversary might be able to learn more about the user than the user
intended to disclose, for example:

 * Cooperating Verifiers might want to track users across services to build
   advertising profiles.
 * Issuers might want to track where users present their credentials to enable
   surveillance.
 * After a data breach at multiple Verifiers, publicly available information
   might allow linking identifiable information presented to Verifier A with
   originally anonymous information presented to Verifier B, therefore revealing
   the identities of users of Verifier B.

The following types of unlinkability are considered here:

 * Verifier/Verifier Unlinkability: Two colluding Verifiers should not be able to
   learn that they have received presentations from the same user.

>  * Issuer/Verifier Unlinkability (Honest Verifier): An Issuer of a credential
>    should not be able to know that a user presented the credential to a certain
>    Verifier (that is behaving according to protocol).
>  * Issuer/Verifier Unlinkability (Colluding/Compromised Verifier): An Issuer of a
>    credential should not be able to tell that a user presented the credential to
>    a certain Verifier (even if the Verifier colludes with the Issuer or becomes
>    compromised and leaks stored credentials from presentations).

Author: ```suggestion
 * Issuer/Verifier Unlinkability (Honest Verifier): An Issuer of a credential
   should not be able to know that a user presented the credential to a certain
   Verifier that is behaving according to protocol.
 * Issuer/Verifier Unlinkability (Colluding/Compromised Verifier): An Issuer of a
   credential should not be able to tell that a user presented the credential to
   a certain Verifier, even if the Verifier colludes with the Issuer or becomes
   compromised and leaks stored credentials from presentations.
```

In all cases, unlinkability is limited to cases where the disclosed claims do
not contain information that directly or indirectly identifies the user. For

> example, when a tax ID is contained in the disclosed claims, the Issuer and

Reviewer 5: ```suggestion
example, when a taxpayer identification number is contained in the disclosed claims, the Issuer and
```

Verifier can easily link the user's transactions. However, when the user only
discloses a birthdate to one Verifier and a postal code to another Verifier, the two Verifiers should not be able to determine that they were interacting with the same user.

Issuer/Verifier unlinkability with a colluding or compromised Verifier cannot be
achieved in salted-hash based approaches to credentials, such as SD-JWT, as the
issued credential is directly forwarded to the Verifier, who can forward it to
the malicious Issuer.

> Issuer/Verifier unlinkability with an honest Verifier can be achieved. A
> 'phone-home' from the Verifier to the Issuer, for example, for a revocation
> check, SHOULD be avoided, or, if that is not possible, MUST be done in a
> privacy-preserving way where details about the credential are not revealed to
> the Issuer. (Even then, timing of such requests might remain as a side-channel.)

Author: How about this?
```suggestion
Contrary to that, Issuer/Verifier unlinkability with an honest Verifier can generally be achieved.
However, a callback from the Verifier to the Issuer, for example, for a revocation
check, can reveal information about the use of the credential to the Verifier. Therefore,
such callbacks SHOULD be avoided, or, if that is not possible, MUST be done in a
privacy-preserving way where details about the credential are not revealed to
the Issuer. (Even then, timing of such requests might remain as a side-channel.)
```

Author: @peppelinux

Reviewer 5: looks good, and I would also contribute with the followings

`However, a callback from the Verifier to the Issuer, for example, for a revocation
check, can reveal information about the use of the credential to the Verifier.` -> `A callback from the Verifier to the Issuer, such as a revocation check, could potentially disclose information about the credential's usage to the Verifier.`

and also

`Therefore,
such callbacks SHOULD be avoided, or, if that is not possible, MUST be done in a
privacy-preserving way where details about the credential are not revealed to
the Issuer. (Even then, timing of such requests might remain as a side-channel.)` -> `Therefore, such callbacks SHOULD be avoided. If they are unavoidable, they MUST be executed in a manner that preserves privacy and does not disclose details about the credential to the Issuer. However, it's important to note that the timing of such requests could potentially serve as a side-channel.`

Author: ```suggestion
Contrary to that, Issuer/Verifier unlinkability with an honest Verifier can generally be achieved.
However, a callback from the Verifier to the Issuer, such as a revocation check, could potentially
disclose information about the credential's usage to the Verifier. Therefore, such callbacks
SHOULD be avoided. Where they are unavoidable, they MUST be executed in a manner that
preserves privacy and does not disclose details about the credential to the Issuer. It is
important to note that the timing of such requests could potentially serve as a side-channel.
```

## Unlinkability

Unlinkability is a property whereby adversaries are prevented from correlating
credential presentations of the same user beyond the user's consent.

> Otherwise, an adversary might be able to learn more about the user than the user
> intended to disclose, for example:

Reviewer: ```suggestion
Without unlinkability, an adversary might be able to learn more about the user than the user
intended to disclose, for example:
```

 * Cooperating Verifiers might want to track users across services to build
   advertising profiles.
 * Issuers might want to track where users present their credentials to enable
   surveillance.
 * After a data breach at multiple Verifiers, publicly available information
   might allow linking identifiable information presented to Verifier A with
   originally anonymous information presented to Verifier B, therefore revealing
   the identities of users of Verifier B.

The following types of unlinkability are considered here:

>  * Verifier/Verifier Unlinkability: Two colluding Verifiers should not be able to

Reviewer: during OID4VCI privacy considerations discussion @jogu talked about the same verifier correlating the user across multiple presentations.

Author: This is solved now.

   learn that they have received presentations from the same user.

>  * Issuer/Verifier Unlinkability (Honest Verifier): An Issuer of a credential
>    should not be able to know that a user presented the credential to a certain
>    Verifier that is behaving according to protocol.

Reviewer: ```suggestion
 * Issuer/Verifier Unlinkability (Honest Verifier): An Issuer of a credential
   should not be able to know that a user presented the credential to a certain
   Verifier that is behaving according to a protocol.
```
what is meant by protocol, here..?

Author: ```suggestion
 * Issuer/Verifier Unlinkability (Honest Verifier): An Issuer of a credential
   should not be able to know that a user presented the credential to a certain
   Verifier that is not behaving maliciously.
```

 * Issuer/Verifier Unlinkability (Colluding/Compromised Verifier): An Issuer of a
   credential should not be able to tell that a user presented the credential to
   a certain Verifier, even if the Verifier colludes with the Issuer or becomes
   compromised and leaks stored credentials from presentations.

In all cases, unlinkability is limited to cases where the disclosed claims do
not contain information that directly or indirectly identifies the user. For
example, when a taxpayer identification number is contained in the disclosed claims, the Issuer and
Verifier can easily link the user's transactions. However, when the user only
discloses a birthdate to one Verifier and a postal code to another Verifier, the two Verifiers should not be able to determine that they were interacting with the same user.

> Issuer/Verifier unlinkability with a colluding or compromised Verifier cannot be
> achieved in salted-hash based approaches to credentials, such as SD-JWT, as the
> issued credential is directly forwarded to the Verifier, who can forward it to
> the malicious Issuer.

Reviewer: ```suggestion
Issuer/Verifier unlinkability with a colluding or compromised Verifier cannot be
achieved in salted-hash based selective disclosure approaches to credentials, such as SD-JWT, as the
issued credential with the Issuer's signature is directly forwarded to the Verifier, who can forward it to
the Issuer, if the Issuer is malicious.
```

forwarding to the malicious Issuer part felt a little abrupt on the first read.

> issued credential with the Issuer's signature is directly forwarded to the Verifier, who can forward it to

Reviewer: ```suggestion
issued credential with the Issuer's signature is directly presented to the Verifier, who can forward it to
```
the first `directly presented/forwarded` is upon user consent from the wallet to the verifier, right? if so, presented feels more natural than forwarding..?

Contrary to that, Issuer/Verifier unlinkability with an honest Verifier can generally be achieved.
However, a callback from the Verifier to the Issuer, such as a revocation check, could potentially

> disclose information about the credential's usage to the Issuer. Therefore, such callbacks
> SHOULD be avoided. Where they are unavoidable, they MUST be executed in a manner that

Reviewer: I am not sure I am ok with `SHOULD be avoided`, given that there are ways to abstract from the issuer who is calling its status endpoint for example.

```suggestion
disclose information about the credential's usage to the Issuer.
Where such callbacks are necessary, they MUST be executed in a manner that
```

> Contrary to that, Issuer/Verifier unlinkability with an honest Verifier can generally be achieved.
> However, a callback from the Verifier to the Issuer, such as a revocation check, could potentially
> disclose information about the credential's usage to the Verifier. Therefore, such callbacks
> SHOULD be avoided. Where they are unavoidable, they MUST be executed in a manner that
> preserves privacy and does not disclose details about the credential to the Issuer. It is
> important to note that the timing of such requests could potentially serve as a side-channel.

Reviewer: ```suggestion
Contrary to that, Issuer/Verifier unlinkability with an honest Verifier can generally be achieved.
However, a callback from the Verifier to the Issuer, such as a revocation check, could potentially
disclose information about the credential's usage to the Issuer. Therefore, such callbacks
SHOULD be avoided. Where they are unavoidable, they MUST be executed in a manner that
preserves privacy and does not disclose details about the credential to the Issuer. It is
important to note that the timing of such requests could potentially serve as a side-channel.
```

## Unlinkability

> Unlinkability is a property that prevents an adversary from correlating
> credential presentations of the same user (beyond the user's consent).

Reviewer 3: This sentence doesn't feel quite right/natural to me. Though I'm not sure this suggestion is particularly great either: 

```suggestion
Unlinkability is a property whereby adversaries are prevented from correlating
credential presentations of the same user (beyond the user's consent).
```

Otherwise, an adversary might be able to learn more about the user than the user
intended to disclose, for example:

 * Cooperating Verifiers might want to track users across services to build
   advertising profiles.
 * Issuers might want to track where users present their credentials to enable
   surveillance.
 * After a data breach at multiple Verifiers, publicly available information
   might allow linking identifiable information presented to Verifier A with
   originally anonymous information presented to Verifier B, therefore revealing
   the identities of users of Verifier B.

The following types of unlinkability are considered here:

 * Verifier/Verifier Unlinkability: Two colluding Verifiers should not be able to
   learn that they have received presentations from the same user.

>  * Issuer/Verifier Unlinkability (Honest Verifier): An Issuer of a credential
>    should not be able to tell that a user presented the credential to a certain
>    Verifier (that is behaving according to protocol).

Reviewer 5: ```suggestion
 * Issuer/Verifier Unlinkability (Honest Verifier): An Issuer of a credential
   should not be able to know that a user presented the credential to a certain
   Verifier (that is behaving according to protocol).
```

 * Issuer/Verifier Unlinkability (Colluding/Compromised Verifier): An Issuer of a
   credential should not be able to tell that a user presented the credential to
   a certain Verifier (even if the Verifier colludes with the Issuer or becomes

>    compromised and leaks stored credentials from presentations).

Reviewer 5: @danielfett this is quite similar, or equivalent, to the previous one. If the concrete information is in the brackets probably we might try a rewording without the brackets, WDYT?

Author: You're right, let's try to remove the brackets.

> In all cases, unlinkability is limited to cases where the disclosed claims do
> not contain information that directly or indirectly identifies the user. For

Reviewer 5: ```suggestion
Unlinkability applies only when the disclosed claims do not hold any information that could directly or indirectly lead to the identification of the user. For
```

Author: I like the 'limited' wording a bit more, it shows that there is an upper bound.

example, when a tax ID is contained in the disclosed claims, the Issuer and
Verifier can easily link the user's transactions. However, when the user only

> discloses a birthdate to one Verifier and a postal code to another Verifier, the
> two Verifiers should not be able to tell that they were talking to the same
> user.

Reviewer 5: ```suggestion
discloses a birthdate to one Verifier and a postal code to another Verifier, the two Verifiers should not be able to determine that they were interacting with the same user.
```

Issuer/Verifier unlinkability with a colluding or compromised Verifier cannot be
achieved in salted-hash based approaches to credentials, such as SD-JWT, as the
issued credential is directly forwarded to the Verifier, who can forward it to
the malicious Issuer.

Issuer/Verifier unlinkability with an honest Verifier can be achieved. A
'phone-home' from the Verifier to the Issuer, for example, for a revocation
check, SHOULD be avoided, or, if that is not possible, MUST be done in a
privacy-preserving way where details about the credential are not revealed to

> the Issuer. (Even then, timing of such requests might remain as a side-channel.)

Reviewer 5: I didn't fully get this, or probably we might try to simplify this text

example, when a taxpayer identification number is contained in the disclosed claims, the Issuer and
Verifier can easily link the user's transactions. However, when the user only
discloses a birthdate to one Verifier and a postal code to another Verifier, the two Verifiers should not be able to determine that they were interacting with the same user.

Issuer/Verifier unlinkability with a colluding or compromised Verifier cannot be
achieved in salted-hash based selective disclosure approaches, such as SD-JWT, as the
issued credential with the Issuer's signature is directly presented to the Verifier, who can forward it to
the Issuer.

Contrary to that, Issuer/Verifier unlinkability with an honest Verifier can generally be achieved.
However, a callback from the Verifier to the Issuer, such as a revocation check, could potentially
disclose information about the credential's usage to the Issuer.
Where such callbacks are necessary, they MUST be executed in a manner that
preserves privacy and does not disclose details about the credential to the Issuer. It is
important to note that the timing of such requests could potentially serve as a side-channel.

> Verifier/Verifier unlinkablility can be achieved using batch issuance: A batch

Reviewer 3: ```suggestion
Verifier/Verifier unlinkablility and presentation unlinkablility can be achieved using batch issuance: A batch
```

Now that https://github.com/oauth-wg/oauth-selective-disclosure-jwt/pull/354/commits/a1cbe276fcd1b5d85ac00a8985e3fd0fb7039f13 introduced presentation unlinkability and the flow of this whole section (where the unlinkablility types are first described and then discussed) it seems like it warrants a mention here. WDYT?
